backends/ubpf/ubpfTable.cpp

```diff
@@ -0,0 +1,408 @@
+            }
+        };  // UbpfActionTranslationVisitor
+    }  // namespace
+
+    void UBPFTableBase::emitInstance(EBPF::CodeBuilder *builder) {
+        builder->append("struct ");
+        builder->appendFormat("ubpf_map_def %s = ", dataMapName);
+        builder->spc();
+        builder->blockStart();
+
+        builder->emitIndent();
+        builder->append(".type = UBPF_MAP_TYPE_HASHMAP,");
+        builder->newline();
+
+        builder->emitIndent();
+        if (keyType->is<IR::Type_Bits>()) {
+            auto tb = keyType->to<IR::Type_Bits>();
+            auto scalar = new UBPFScalarType(tb);
+            builder->append(".key_size = sizeof(");
+            scalar->emit(builder);
+            builder->append("),");
+        } else {
+            builder->appendFormat(".key_size = sizeof(struct %s),",
+                                  keyTypeName.c_str());
+        }
+        builder->newline();
+
+        builder->emitIndent();
+        if (valueType->is<IR::Type_Bits>()) {
+            auto tb = valueType->to<IR::Type_Bits>();
+            auto scalar = new UBPFScalarType(tb);
+            builder->append(".value_size = sizeof(");
+            scalar->emit(builder);
+            builder->append("),");
+        } else {
+            builder->appendFormat(".value_size = sizeof(struct %s),",
+                                  valueTypeName.c_str());
+        }
+        builder->newline();
+
+        builder->emitIndent();
+        builder->appendFormat(".max_entries = %d,", size);
+        builder->newline();
+
+        builder->emitIndent();
+        builder->append(".nb_hash_functions = 0,");
+        builder->newline();
+
+        builder->blockEnd(false);
+        builder->endOfStatement(true);
+    }
+
+    UBPFTable::UBPFTable(const UBPFProgram *program,
+                         const IR::TableBlock *table,
+                         EBPF::CodeGenInspector *codeGen) :
+            UBPFTableBase(program, EBPFObject::externalName(table->container),
+                          codeGen), table(table) {
+
+        cstring base = instanceName + "_defaultAction";
+        defaultActionMapName = program->refMap->newName(base);
+
+        base = table->container->name.name + "_actions";
+        actionEnumName = program->refMap->newName(base);
+
+        base = instanceName + "_NoAction";
+        noActionName = program->refMap->newName(base);
+
+        keyGenerator = table->container->getKey();
+        actionList = table->container->getActionList();
+
+        setTableSize(table);
+    }
+
+    void UBPFTable::setTableSize(const IR::TableBlock *table) {
+        auto properties = table->container->properties->properties;
+        size = UINT16_MAX; // Default value 2^16. Next power is too big for ubpf vm.
+                           // For instance, 2^17 causes error while loading program.
+        for (auto property : properties) {
+            if (property->name.name == table->container->properties->sizePropertyName &&
```

Comment:
Maybe you can use the `getConstantProperty` method of P4::Table

Reply:
Refactored.